### PR TITLE
Use spans

### PR DESCRIPTION
**Pull request by tabrath**
All bases except base58. Much room for improvements and optimizations!

### CODE REVIEW

File: src/Multiformats.Base/Base10.cs [hunk 1]
```diff
@@ -13,26 +12,68 @@ internal class Base10 : Multibase
...
+            zeroLeaded[1] = '0';
+            input.CopyTo(zeroLeaded.Slice(2));
+            var big = BigInteger.Parse(zeroLeaded.ToString(), NumberStyles.None);
+
+            var leadingZeros = LeadingZeros(input);
+            var bigBytes = big.ToByteArray().AsSpan();
+            bigBytes.Reverse();
+            var bigLeadingZeros = LeadingZeros(bigBytes);
+
+            Span<byte> result = new byte[leadingZeros + (bigBytes.Length - bigLeadingZeros)];
```
Comment (hanabi1224) [inline src/Multiformats.Base/Base10.cs]: I believe this allocates heap memory, and does not reduce GC overhead

**Comment by hanabi1224** [comment]
I notice there're some Span<byte> foo = new byte[] code, unless it's intentional to avoid stackoverflow, I believe it still allocate heap memory thus does not reduce GC overhead.

I created a simple benchmark
```csharp
    [CoreJob]
    [MemoryDiagnoser, GcServer(true)]
    [RPlotExporter, RankColumn]
    public class ToyTask
    {
        [Params(1024, 10240)]
        public int size;

        [Benchmark]
        public ReadOnlyMemory<byte> Span_New()
        {
            Span<byte> span = new byte[size];
            Fill(span, 88);
            return span.ToArray();
        }

        [Benchmark]
        public ReadOnlyMemory<byte> Span_Stackalloc()
        {
            Span<byte> span = stackalloc byte[size];
            Fill(span, 88);
            return span.ToArray();
        }

        private static void Fill(Span<byte> span, byte value)
        {
            for (var i = 0; i < span.Length; i++)
            {
                span[i] = value;
            }
        }
    }
```

And got
``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-7600U CPU 2.80GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.2.300
  [Host]     : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), 64bit RyuJIT
  Job-RWMZYW : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), 64bit RyuJIT

Runtime=Core  Server=True  

```
|          Method |  size |       Mean |      Error |     StdDev |     Median | Rank |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|---------------- |------ |-----------:|-----------:|-----------:|-----------:|-----:|-------:|-------:|------:|----------:|
|        **Span_New** |  **1024** |   **691.4 ns** |  **18.262 ns** |  **50.604 ns** |   **674.9 ns** |    **1** | **0.0887** |      **-** |     **-** |   **2.05 KB** |
| Span_Stackalloc |  1024 |   669.8 ns |   8.471 ns |   7.074 ns |   667.2 ns |    1 | 0.0429 |      - |     - |   1.02 KB |
|        **Span_New** | **10240** | **6,762.3 ns** | **400.168 ns** | **506.085 ns** | **6,598.4 ns** |    **2** | **0.8316** | **0.0153** |     **-** |  **20.05 KB** |
| Span_Stackalloc | 10240 | 6,854.9 ns |  87.995 ns |  78.005 ns | 6,831.5 ns |    3 | 0.4120 |      - |     - |  10.02 KB |

**Comment by tabrath** [comment]
Does that mean we should return ReadOnlyMemory instead of ReadOnlySpan for the public methods?

**Comment by hanabi1224** [comment]
@tabrath I believe a function is not allowed to return stack memory in c#, instead the function should take Span<T> as parameter and do manipulations on it directly.  So functions in system library use parameter as appose to return value to allow no GC allocation operations (it's caller's responsibility to decide heap, stack or even unmanaged memory to use)  e.g. [BigInteger.TryWriteBytes(Span<Byte>, Int32, Boolean, Boolean)](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.trywritebytes?view=netcore-2.2#System_Numerics_BigInteger_TryWriteBytes_System_Span_System_Byte__System_Int32__System_Boolean_System_Boolean_)

To summarize my understanding of c# span, it has the limitation that, if ppl want to use no unsafe code and avoid gc overhead, he has to use stack memory.  In other word, even with span, it's not possible to use heap memory with no GC and no unsafe code

image

**Comment by tabrath** [comment]
@hanabi1224 Maybe we should consider a public api where we leave the allocation up to the user, like `TryEncode` and `TryDecode`?

**Comment by hanabi1224** [comment]
@tabrath That will be great, but it's hard for ppl to know the memory size to allocate, BigInterger has a [GetByteCount](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.getbytecount?view=netcore-2.2) method to facilitate that.  However, that sounds like a larger change to provide api to calculate buffer size.  IMHO, with current api, there will be still a lot of gain if we reduce GC overhead in function's internal implementation.  (My intuition is that Memory / ReadOnlyMemory is better return type than Span but I have not thought through)

**Comment by rvagg** [comment]
closing due to inactivity, archiving repo
